src/deeponet.jl

```diff
@@ -0,0 +1,116 @@
+
+## Example:
+```julia-repl
+julia> deeponet = DeepONet(; branch = (64, 32, 32, 16), trunk = (1, 8, 8, 16))
+
+Branch : Chain(
+    layer_1 = Dense(64 => 32),          # 2_080 parameters
+    layer_2 = Dense(32 => 32),          # 1_056 parameters
+    layer_3 = Dense(32 => 16),          # 528 parameters
+)         # Total: 3_664 parameters,
+          #        plus 0 states.
+ 
+  Trunk: Chain(
+    layer_1 = Dense(1 => 8),            # 16 parameters
+    layer_2 = Dense(8 => 8),            # 72 parameters
+    layer_3 = Dense(8 => 16),           # 144 parameters
+)         # Total: 232 parameters,
+          #        plus 0 states.
+()  # 3_896 parameters
+```
+"""
+function DeepONet(; branch = (64, 32, 32, 16), trunk = (1, 8, 8, 16),
+    branch_activation = identity, trunk_activation = identity)
+
+    # checks for last dimension size
+    @assert branch[end] == trunk[end] "Branch and Trunk net must share the same amount of nodes in the last layer. 
+    Otherwise Σᵢ bᵢⱼ tᵢₖ won't work."
+
+    branch_net = Chain([Dense(branch[i] => branch[i+1], branch_activation)
+    for i in 1:length(branch) - 1]...);
+
+    trunk_net = Chain([Dense(trunk[i] => trunk[i+1], trunk_activation)
+    for i in 1:length(trunk) - 1]...);
+
+    return DeepONet(branch_net, trunk_net);
+end
+
+"""
+    DeepONet(branch::L1, trunk::L2) where {L1, L2}
+constructs a DeepONet from a `branch` and `trunk` architectures. 
+Make sure that both the nets output should have the same first dimension.
+
+## Arguments:
+- `branch`: `Lux` network to be used as branch net. 
+- `trunk`: `Lux` network to be used as trunk net. 
+
+
+## References:
+Lu Lu, Pengzhan Jin, George Em Karniadakis , "DeepONet: Learning nonlinear operators for identifying differential equations based on the universal approximation theorem of operators"
+doi: https://arxiv.org/abs/1910.03193
+
+## Example:
+
+```julia-repl
+julia> branch_net =  Chain(Dense(64 =>32), Dense(32 =>32), Dense(32 => 16));
+julia> trunk_net = Chain(Dense(1 =>8), Dense(8 =>8), Dense(8 => 16));
+julia> don_ = DeepONet(branch_net, trunk_net)
+
+Branch : Chain(
+    layer_1 = Dense(64 => 32),          # 2_080 parameters        
+    layer_2 = Dense(32 => 32),          # 1_056 parameters        
+    layer_3 = Dense(32 => 16),          # 528 parameters
+)         # Total: 3_664 parameters,
+          #        plus 0 states.
+
+  Trunk: Chain(
+    layer_1 = Dense(1 => 8),            # 16 parameters
+    layer_2 = Dense(8 => 8),            # 72 parameters
+    layer_3 = Dense(8 => 16),           # 144 parameters
+)         # Total: 232 parameters,
+          #        plus 0 states.
+()  # 3_896 parameters
+```
+"""
+function DeepONet(branch::L1, trunk::L2) where {L1, L2}
+
+    io = IOBuffer();
+    td= TextDisplay(io);
+    display(td, branch);
+    branch_name = String(take!(io));
+
+    display(td, trunk);
+    trunk_name = String(take!(io));
+
+    name = "Branch : $branch_name \n  Trunk: $trunk_name"
```

Comment:
Let the original printing be

Reply:
The original printing would also print the `@assert` block and didn't look very tidy. Comes out as:
```
DeepONet(branch_net, trunk_net)
@compact(
    branch = Chain(
        layer_1 = Dense(64 => 32),      # 2_080 parameters        
        layer_2 = Dense(32 => 32),      # 1_056 parameters        
        layer_3 = Dense(32 => 16),      # 528 parameters
    ),
    trunk = Chain(
        layer_1 = Dense(1 => 8),        # 16 parameters
        layer_2 = Dense(8 => 8),        # 72 parameters
        layer_3 = Dense(8 => 16),       # 144 parameters
    ),
) do (u, y)
    t = trunk(y)
    b = branch(u)
    #= C:\Users\ayush\Desktop\pkgs\LuxNeuralOperators.jl\src\deeponet.jl:28 =# @assert size(t, 1) == size(b, 1) "Branch and Trunk net must share the same amount of nodes in the last layer. \n        Otherwise Σᵢ bᵢⱼ tᵢₖ won't work."
    tᵀ = permutedims(t, (2, 1, 3))
    b_ = permutedims(reshape(b, size(b)..., 1), (1, 3, 2))        
    G = batched_mul(tᵀ, b_)
    return dropdims(G; dims = 2)
end       # Total: 3_896 parameters,
          #        plus 0 states.
```

Reply:
I see, then overload something like https://github.com/LuxDL/Lux.jl/blob/f7f45512c1240d24e76f1bfcbb3bd26eba351e65/src/helpers/stateful.jl#L55

```julia
Base.show(io::IO, s::CompactLuxLayer{:DeepONet})
```

Reply:
I had initially wanted to do this, but I won't know inside the function what architectures `branch` and `trunk` would have. @avik-pal am I missing something?

Reply:
They are available from `<model_name>.layers.<branch|trunk>`

Reply:
The `model_name.layers` is an empty NamedTuple here.
```
don = DeepONet(
           (64, 32, 32, 16), (1, 8, 8, 16)
       )
don.layers
```
returns 
`NamedTuple()`

Reply:
```julia-repl
julia> model = DeepONet()
Branch : Chain(
    layer_1 = Dense(64 => 32),          # 2_080 parameters
    layer_2 = Dense(32 => 32),          # 1_056 parameters
    layer_3 = Dense(32 => 16),          # 528 parameters
)         # Total: 3_664 parameters,
          #        plus 0 states.
 
  Trunk: Chain(
    layer_1 = Dense(1 => 8),            # 16 parameters
    layer_2 = Dense(8 => 8),            # 72 parameters
    layer_3 = Dense(8 => 16),           # 144 parameters
)         # Total: 232 parameters,
          #        plus 0 states.
()  # 3_896 parameters

julia> model.layers
(branch = Chain(), trunk = Chain())
```

Pull in the changes from `clean_deeponet` branch. Also set your config to rebase instead of merging `git config --global pull.rebase true`